### PR TITLE
Fix forceful creation of temporary tables

### LINKED ISSUE #42525: Creation of temporary table with `force: true` issues a non-temporary table drop when using the MySQL adapter
### Steps to reproduce
Create a vanilla Rails 6.1.3.2 application using the `mysql2` adapter and execute the following command inside a `rails console`:

```ruby
ActiveRecord::Base.connection.create_table(:foo, force: true, temporary: true) { |t| t.string :name }
```

### Expected behavior

The drop statement should include the `TEMPORARY` keyword.

```sql
DROP TEMPORARY TABLE IF EXISTS `foo`
CREATE TEMPORARY TABLE `foo` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `name` varchar(255))
```

### Actual behavior

The drop statement does not include the `TEMPORARY` keyword.
```sql
DROP TABLE IF EXISTS `foo`
CREATE TEMPORARY TABLE `foo` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `name` varchar(255))
```

### Investigation details

This bug was found due to the following error reported by a failing spec that was trying to create a temporary table with `force: true` in an rspec's `before` block.
```
ActiveRecord::StatementInvalid:
  Mysql2::Error: SAVEPOINT active_record_1 does not exist
```

From [ActiveRecord::Transactions documentation](https://api.rubyonrails.org/v6.1.3.2/classes/ActiveRecord/Transactions/ClassMethods.html#module-ActiveRecord::Transactions::ClassMethods-label-Caveats):
> If you're on MySQL, then do not use Data Definition Language (DDL) operations in nested transactions blocks that are emulated with savepoints. That is, do not execute statements like 'CREATE TABLE' inside such blocks. This is because MySQL automatically releases all savepoints upon executing a DDL operation. When transaction is finished and tries to release the savepoint it created earlier, a database error will occur because the savepoint has already been automatically released

The [drop operation for non-temporary tables commits the current transaction](https://mariadb.com/kb/en/drop-table/) in MariaDB:
> Note: DROP TABLE automatically commits the current active transaction, unless you use the TEMPORARY keyword.

The following part in version 6.1.3.2 of [ActiveRecord::ConnectionAdapters::SchemaStatements](https://github.com/rails/rails/blob/v6.1.3.2/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L316-L317)
```ruby
if force
  drop_table(table_name, force: force, if_exists: true)
else
  schema_cache.clear_data_source_cache!(table_name.to_s)
end
```
does not pass the `temporary` option in the `drop_table` call and thus the [ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter](https://github.com/rails/rails/blob/v6.1.3.2/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L334-L337)
```ruby
def drop_table(table_name, **options)
  schema_cache.clear_data_source_cache!(table_name.to_s)
  execute "DROP#{' TEMPORARY' if options[:temporary]} TABLE#{' IF EXISTS' if options[:if_exists]} #{quote_table_name(table_name)}#{' CASCADE' if options[:force] == :cascade}"
end
```
does not include the `TEMPORARY` keyword in the produced SQL.

### System configuration
**Rails version**: 6.1.3.2

**Ruby version**: 2.7.3

### PR DESCRIPTION
**Pull request by iridakos**
### Summary

As documented in #42525 in Rails 6.1.3.2, the following code:

```ruby
ActiveRecord::Base.connection.create_table(:foo, force: true, temporary: true) { |t| t.string :name }
```

is issuing a non-temporary table drop with the same name when using the mysql2 adapter:

```sql
DROP TABLE IF EXISTS `foo`
CREATE TEMPORARY TABLE `foo` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `name` varchar(255))
```

This PR fixes the issue by properly passing the `temporary` option (initially provided to the `create_table` call) to the `drop_table` call of the underlying adapter.

### Notes 

1. I've added a related test in `activerecord/test/cases/adapters/mysql2/active_schema_test.rb` but if you believe that there's a better place to host it let me know and I will proceed to the required changes.

2. In the test, I initially wrapped the table creation call in an AR transaction so that the following error would be raised without the fix
  ```ruby
  ActiveRecord::StatementInvalid:
    Mysql2::Error: SAVEPOINT active_record_1 does not exist
  ```
  but reading the rest of the specs, I decided that using the `assert_sql` to check the generated SQL statements would be a better fit. What do you think?

---

Fixes #42525

### CODE REVIEW

File: activerecord/test/cases/adapters/mysql2/active_schema_test.rb [hunk 1]
```diff
@@ -180,6 +180,19 @@ def test_indexes_in_create
     assert_match expected, actual
   end
 
+  def test_forceful_temporary_table_creation
```
Comment (simi) [inline activerecord/test/cases/adapters/mysql2/active_schema_test.rb:183]: Since this is not related to `mysql2` adapter directly, I think it would fit https://github.com/rails/rails/blob/main/activerecord/test/cases/migration/change_schema_test.rb test file better (including already various `create_table `tests).

File: activerecord/test/cases/adapters/mysql2/active_schema_test.rb [hunk 2]
```diff
@@ -180,6 +180,19 @@ def test_indexes_in_create
...
+    expected_sql = [
+      /\ADROP TEMPORARY TABLE IF EXISTS `forced_temp`/,
+      /\ACREATE TEMPORARY TABLE `forced_temp` \(`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY\)/
+    ]
+
+    assert_sql(*expected_sql) do
+      ActiveRecord::Base.connection.create_table(:forced_temp, temporary: true, force: true)
+    end
+  ensure
+    ActiveRecord::Base.connection.drop_table :forced_temp, temporary: true rescue nil
```
Comment (simi) [inline activerecord/test/cases/adapters/mysql2/active_schema_test.rb:193]: I think you can use `if_exists: true` to ensure it is dropped only when present instead of `rescue nil` which can swallow other exceptions. Or is there any other reason to use `rescue nil` here?

File: activerecord/test/cases/adapters/mysql2/active_schema_test.rb (hunk 1 above)
Comment (iridakos) [reply activerecord/test/cases/adapters/mysql2/active_schema_test.rb:183]: @simi Good point, thank you! I will add it there.

File: activerecord/test/cases/adapters/mysql2/active_schema_test.rb (hunk 2 above)
Comment (iridakos) [reply activerecord/test/cases/adapters/mysql2/active_schema_test.rb:193]: I followed the approach used elsewhere in this file for consistency, but I think you're right.

```ruby
ActiveRecord::Base.connection.drop_table :delete_me rescue nil
```

File: activerecord/test/cases/adapters/mysql2/active_schema_test.rb (hunk 1 above)
Comment (iridakos) [reply activerecord/test/cases/adapters/mysql2/active_schema_test.rb]: I would like some help here.

- The target [test file](https://github.com/rails/rails/blob/main/activerecord/test/cases/migration/change_schema_test.rb) doesn't have checks for the produced SQL but only assertions on the newly created table whereas in my case I need to check that the `temporary` directive is being added to the query. Is it OK to introduce such checks there?

- I suppose I should add specs also for the rest of the adapters (currently I check only against mysql, sqlite & postgresql). Could someone point me on the documentation on how to setup the suite for the other adapters as well?
